engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java

```diff
@@ -51,6 +54,25 @@ public InputStream inputStream(OpenOption[] opts) throws IOException {
     return this.truffleFile.newInputStream(opts);
   }
 
+  @Builtin.Method(name = "read_last_bytes_builtin")
+  @Builtin.WrapException(from = IOException.class, to = PolyglotError.class, propagate = true)
+  public Array readLastBytes(long n) throws IOException {
+    try (SeekableByteChannel channel = this.truffleFile.newByteChannel(Set.of(StandardOpenOption.READ))) {
+      int bytesToRead = Math.toIntExact(Math.min(channel.size(), n));
+      channel.position(channel.size() - bytesToRead);
+      ByteBuffer buffer = ByteBuffer.allocate(bytesToRead);
+      while (buffer.hasRemaining()) {
+        channel.read(buffer);
+      }
+
+      Object[] result = new Long[bytesToRead];
+      for (int i = 0; i < bytesToRead; ++i) {
+        result[i] = (long) buffer.get(i);
+      }
+      return new Array(result);
+    }
+  }
```

Comment:
I have to admit this feels slightly sketchy - writing a very specific function as a builtin.

I guess a more general way would be to create a method that returns the `RandomAccessFile` in a similar manner to `InputStream` and then have an Enso wrapper similar to `Input_Stream`. However this seemed like an overkill for such a simple problem and when I was first implementing this I wasn't ever sure yet if we will use this finally.

I can rewrite it to create some simple `Random_Access_File` Enso wrapper if you think it is necessary, although admittedly all this buffer work seems a bit simpler to do in Java.

Reply:
I think as this is an internal method for now happy to leave as is - if we find a need for random access more generally we can add it.